src/python/library/tritonclient/utils/__init__.py

```diff
@@ -269,3 +269,76 @@ def deserialize_bytes_tensor(encoded_tensor):
         offset += l
         strs.append(sb)
     return (np.array(strs, dtype=np.object_))
+
+
+def serialize_bf16_tensor(input_tensor):
+    """
+    Serializes a bfloat16 tensor into a flat numpy array of bytes.
+    The numpy array should use dtype of np.float32.
+
+    Parameters
+    ----------
+    input_tensor : np.array
+        The bfloat16 tensor to serialize.
+
+    Returns
+    -------
+    serialized_bf16_tensor : np.array
+        The 1-D numpy array of type uint8 containing the serialized bytes in 'C' order.
+
+    Raises
+    ------
+    InferenceServerException
+        If unable to serialize the given tensor.
+    """
+
+    if input_tensor.size == 0:
+        return np.empty([0], dtype=np.object_)
+
+    # If the input is a tensor of float32, then must flatten those into
+    # a 1-dimensional array containing the element bytes. All elements
+    # are concatenated together in "C" order.
+
+    if (input_tensor.dtype == np.float32):
+        flattened_ls = []
+        for obj in np.nditer(input_tensor, flags=["refs_ok"], order='C'):
+            # To trunctate the float32 to a bfloat16, we need the high-order bits.
+            obj_bytes = struct.pack("<f", obj)[2:4]
+            flattened_ls.append(obj_bytes)
+        flattened = b''.join(flattened_ls)
+        flattened_array = np.asarray(flattened, dtype=np.object_)
+        if not flattened_array.flags['C_CONTIGUOUS']:
```

Comment:
This may be my lack of knowledge in python, but why do we need to do: `flattened_ls ->flattened -> flattened_array -> (optional) ascontigiousarray`? Why isn't flattened_ls the bytes we need?

Reply:
I tried to have relative parity with serialize_byte_tensor, since we're treating bfloat16 as bytes, so what follows is my best understanding of why serialize_byte_tensor used this logic:

- Flattened_ls is a list of bytes.
- Then, flattened joins those into one byte string. (Viewing git blame, these two steps were done because they are faster than appending to a byte string directly each time.)
- Flattened_array is creating a Numpy array with object, which is what we need to pass in.
- Flattened_array should be contiguous in C order, since flattened should have been stored in contiguous memory and this is Numpy's default. However, the extra flag seems to be making sure that that's always the case. We could also directly call .ascontiguousarray and get rid of the third call, though we may be doing the flag check for efficiency: https://github.com/numpy/numpy/issues/11804

Let me know if you see something that can be improved. I could make the naming for all of these uniform so that we're not creating new variables, though I imagine the compiler might optimize out the renamings anyway.

Reply:
That makes sense. As I thought, this was my lack of python knowledge.

>Then, flattened joins those into one byte string. (Viewing git blame, these two steps were done because they are faster than appending to a byte string directly each time.)

Makes sense. If you know the length of the array passed in then you can directly allocate the size needed to hold the copy rather than appending and resizing as you go.